pkg/build/controller/buildimagechangetrigger/buildimagechangetrigger_controller.go

```diff
@@ -0,0 +1,287 @@
+	c.isStoreSynced = isInformer.HasSynced
+
+	return c
+}
+
+func (bc *BuildImageChangeTriggerController) addImageStream(obj interface{}) {
+	stream := obj.(*imageapi.ImageStream)
+	glog.V(4).Infof("Imagestream %s/%s added", stream.Namespace, stream.Name)
+	bc.enqueueImageStream(stream)
+}
+
+func (bc *BuildImageChangeTriggerController) updateImageStream(old, cur interface{}) {
+	// if this a periodic relist for all known streams, HandleImageStream will handle checking the
+	// last image ref and ensuring we don't create a duplicate build for the same image stream event;
+	// as such we don't attempt to filter events at this level
+	newStream := cur.(*imageapi.ImageStream)
+
+	glog.V(4).Infof("Imagestream %s/%s updated.", newStream.Name, newStream.Namespace)
+	bc.enqueueImageStream(newStream)
+}
+
+func (bc *BuildImageChangeTriggerController) enqueueImageStream(stream *imageapi.ImageStream) {
+	key, err := kcontroller.KeyFunc(stream)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", stream, err))
+		return
+	}
+	bc.queue.Add(key)
+}
+
+func (c *BuildImageChangeTriggerController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Wait for the bc store to sync before starting any work
+	if !cache.WaitForCacheSync(stopCh, c.bcStoreSynced, c.isStoreSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down build image change trigger controller")
+}
+
+func (bc *BuildImageChangeTriggerController) worker() {
+	for {
+		if quit := bc.work(); quit {
+			return
+		}
+	}
+}
+
+func (bc *BuildImageChangeTriggerController) work() bool {
+	key, quit := bc.queue.Get()
+	if quit {
+		return true
+	}
+	defer bc.queue.Done(key)
+
+	is, err := bc.getImageStreamByKey(key.(string))
+	if err != nil {
+		utilruntime.HandleError(err)
+	}
+
+	if is == nil {
+		return false
+	}
+
+	err = bc.HandleImageStream(is)
+	bc.handleError(err, key, is)
+
+	return false
+
+}
+
+func (bc *BuildImageChangeTriggerController) getImageStreamByKey(key string) (*imageapi.ImageStream, error) {
+	obj, exists, err := bc.isStore.GetByKey(key)
+	if err != nil {
+		glog.V(4).Infof("Unable to retrieve imagestream %q from store: %v", key, err)
+		bc.queue.AddRateLimited(key)
+		return nil, err
+	}
+	if !exists {
+		glog.V(4).Infof("ImageStream %q is no longer in the shared informer store to imagestream lister cache (perhaps it has been deleted)", key)
+		return nil, nil
+	}
+
+	return obj.(*imageapi.ImageStream), nil
+}
+
+func (bc *BuildImageChangeTriggerController) HandleImageStream(stream *imageapi.ImageStream) error {
+	glog.V(4).Infof("Build image change controller detected ImageStream change for %s/%s and docker image respository %s", stream.Namespace, stream.Name, stream.Status.DockerImageRepository)
```

Comment:
this message is misleading in that it may not be a change ... it could be an imagestream was added or simply relisted periodically. I think a simpler "handling imagestream %s/%s with docker image repository %s" would work.